### PR TITLE
Delete service info from cluster when service is disabled

### PR DESCRIPTION
**Pull request by abhi**
This PR contains a fix for moby/moby#30321. There was a moby/moby#31142
PR intending to fix the issue by adding a delay between disabling the
service in the cluster and the shutdown of the tasks. However
disabling the service was not deleting the service info in the cluster.
Added a fix to delete service info from cluster and verified using siege
to ensure there is zero downtime on rolling update of a service.

Signed-off-by: Abhinandan Prativadi <abhi@docker.com>

### CODE REVIEW

File: sandbox.go [hunk 1]
```diff
@@ -691,6 +691,9 @@ func (sb *sandbox) EnableService() error {
 func (sb *sandbox) DisableService() error {
 	logrus.Debugf("DisableService %s START", sb.containerID)
 	for _, ep := range sb.getConnectedEndpoints() {
+		if err := ep.deleteServiceInfoFromCluster(sb, "DisableService"); err != nil {
```
Comment (fcrisciani) [inline sandbox.go]: This logic was intentionally removed to avoid a race with the sbLeave. So far the sbLeave is the only code path that actually cleans the name resolution/load balancer.

Comment (abhinandanpb) [inline sandbox.go]: shouldnt service disabling be independent of sbleave ? Removing the endpoint from a network part of a sandbox must be decoupled from the service disabling. This helps in make before break scenarios. If its tightly coupled then we might end up with issue like the above. Just like enabling service

Comment (sanimej) [inline sandbox.go]: @fcrisciani That explains why the [moby #31142](https://github.com/moby/moby/pull/31142) patch. doesn't help in 17.06. 

A valid use case shouldn't be broken to fix the race. In this case we can avoid the race by checking the `serviceEnabled` flag. IOW, `deleteServiceInfoFromCluster` will be called only once for one endpoint.

Comment (fcrisciani) [inline sandbox.go]: @sanimej I'm not saying is wrong. I'm saying that we have to test it properly for all the other cases that we work in the past week to be sure to not introduce again races here.

Comment (abhinandanpb) [inline sandbox.go]: agreed on that. From what I see in here - I see that service is disabled first before task shutdown. The race condition was seen on the remote nodes when this happens ? Just making sure I incorporate the scenario in my testing.   
With the current code we are basically adding 2 second delay to do nothing so we have increased our convergence time.

File: agent.go [hunk 2]
```diff
@@ -665,6 +665,12 @@ func (ep *endpoint) addServiceInfoToCluster(sb *sandbox) error {
 }
 
 func (ep *endpoint) deleteServiceInfoFromCluster(sb *sandbox, method string) error {
+
+	if !ep.serviceEnabled {
```
Comment (fcrisciani) [inline agent.go]: I don't see the serviceEnable flag being saved into the store in the Marshal of the endpoint, are we sure that does make sense then this value?

File: sandbox.go (hunk 1 above)
Comment (fcrisciani) [inline sandbox.go]: The getConnected endpoints coming from the sandbox should return only the endpoints that are locally deployed. Conceptually should work fine, is it guaranteed serialization between EnableService/DisableService and CreateEndpoint?
Like is it possible that the DisableService is called while a CreateEndpoint happens and so the Endpoint is added to the service after being disabled?

File: agent.go (hunk 2 above)
Comment (abhinandanpb) [inline agent.go]: it should be only part of the local store and in memory right ? This need not be propogate to the store. So Marshalling and Unmarshalling may not be needed ? getendpoints returns endpoints from in memory.

File: sandbox.go (hunk 1 above)
Comment (abhinandanpb) [inline sandbox.go]: getconnected endpoints is syncronised operation. So I wouldnt expect this to change around either of the operations.
Enabling of a service happens after endpoint creation and then the service enabled flag is set (I see a problem in old code. will correct that as well) and Disabling of a service is done on task shutdown before the deleteEndpoint.

File: agent.go (hunk 2 above)
Comment (sanimej) [inline agent.go]: The possibility of concurrent calls too deleteServiceInfoFromCluster is from sb.DisableService (which calls getConnectedEndpoints) and ep.Leave(). The latter fetches the endpoint from the store. So the `serviceEnabled` field has to be marshalled. Also, after setting it ep has to be updated in the store.

File: sandbox.go (hunk 1 above)
Comment (sanimej) [inline sandbox.go]: @abhinandanpb `serviceEnabled` is set before the `addServiceInfoToCluster` call and its reset if the call fails. Can you clarify what the problem in the current code is ?

**Comment by mvdstam** [comment]
@thaJeztah Any chance to get this merged in so we can close [#30321](https://github.com/moby/moby/issues/30321)?

**Comment by thaJeztah** [comment]
@mvdstam I'm not a maintainer in this repository, but I can try asking what the status is on  this one

**Comment by abhinandanpb** [comment]
ping @mavenugo @fcrisciani @sanimej

**Comment by mvdstam** [comment]
@thaJeztah Thanks, sorry for pulling you in on this one. 😉

**Comment by mvdstam** [comment]
@abhinandanpb @fcrisciani Hey guys, any chance you can take a look at this? Getting this merged soon would be really awesome; can't wait to have true zero-downtime deployments with Docker Swarm. 😃 

Thanks!

File: sandbox.go [hunk 3]
```diff
@@ -676,23 +676,68 @@ func (sb *sandbox) SetKey(basePath string) error {
...
+			// enable service on the endpoint copy in the sandbox
+			ep.enableService(true)
+
+			n, err := ep.getNetworkFromStore()
+			if err != nil {
+				logrus.Warnf("could not enable service on sandbox:%s,endpoint:%s,err: %v", sb.ID(), ep.Name(), err)
+				continue
+			}
+
+			store := n.getController().getStore(ep.DataScope())
```
Comment (fcrisciani) [inline sandbox.go]: this looks unnecessary, the same logic is replicated inside the updateToStore()

```
cs := c.getStore(kvObject.DataScope())
	if cs == nil {
		return ErrDataStoreNotInitialized(kvObject.DataScope())
	}
```

File: sandbox.go [hunk 4]
```diff
@@ -676,23 +676,68 @@ func (sb *sandbox) SetKey(basePath string) error {
...
+		}
+
+		if ep.isServiceEnabled() {
+			if err := ep.deleteServiceInfoFromCluster(sb, "DisableService"); err != nil {
+				logrus.Warnf("could not update state for endpoint %s into cluster: %v", ep.Name(), err)
+			}
+
+			ep.enableService(false)
+
+			store := n.getController().getStore(ep.DataScope())
```
Comment (fcrisciani) [inline sandbox.go]: same here, this is done inside the updateToStore()

File: sandbox.go [hunk 5]
```diff
@@ -676,23 +676,68 @@ func (sb *sandbox) SetKey(basePath string) error {
...
+
+			if store == nil {
+				logrus.Warnf("store not found for scope %s on enable service on endpoint:%s", ep.DataScope(), ep.Name())
+			}
+
+			//Write the ep copy to the store
+			ep.enableService(true)
+			if err := n.getController().updateToStore(ep); err != nil {
+				break
```
Comment (fcrisciani) [inline sandbox.go]: also when we break here basically the service got already updated but now the flag will remain at false, so there would not be any disable service happening neither in the case of sbLeave

File: sandbox.go [hunk 6]
```diff
@@ -676,23 +676,68 @@ func (sb *sandbox) SetKey(basePath string) error {
...
-		if ep.enableService(true) {
+		if !ep.isServiceEnabled() {
 			if err := ep.addServiceInfoToCluster(sb); err != nil {
-				ep.enableService(false)
 				return fmt.Errorf("could not update state for endpoint %s into cluster: %v", ep.Name(), err)
 			}
+			// enable service on the endpoint copy in the sandbox
+			ep.enableService(true)
+
+			n, err := ep.getNetworkFromStore()
```
Comment (fcrisciani) [inline sandbox.go]: I would move this on the top before the addServiceInfoToCluster, same reason, if we cannot save the flag then we cannot activate the service

File: sandbox.go [hunk 7]
```diff
@@ -676,23 +676,68 @@ func (sb *sandbox) SetKey(basePath string) error {
...
+
 	for _, ep := range sb.getConnectedEndpoints() {
-		if ep.enableService(true) {
+		if !ep.isServiceEnabled() {
 			if err := ep.addServiceInfoToCluster(sb); err != nil {
-				ep.enableService(false)
 				return fmt.Errorf("could not update state for endpoint %s into cluster: %v", ep.Name(), err)
 			}
+			// enable service on the endpoint copy in the sandbox
+			ep.enableService(true)
```
Comment (abhinandanpb) [inline sandbox.go]: looks like I deleted a line. One was to update the store and one to update in mem copy. Let me rework this and test it out.

**Comment by mvdstam** [comment]
Hey @abhinandanpb, your patch has been confirmed to definitely fix [moby/moby#30321](https://github.com/moby/moby/issues/30321). This means that true zero-downtime deployments with Docker Swarm Mode will finally be possible; this is really awesome! Thanks for your work and hope to see it released shortly. 👍

**Comment by hjdr4** [comment]
Hello guys,

Patch idea is good while current implementation is bad.

I tried to make a proper fix for this based on the discussions: https://github.com/hjdr4/libnetwork/commit/8bc2fbc87b3ec2ed830280cf91e92ce1a16fec7a

I don't want to put mess in the process but I would appreciate this bug is fixed because this is show stopper for production usage of Swarm. 17.06.1 is on the way, having the fix merged would be nice (I may be dreaming, I don't know what can go or not in minor updates).

I hope that helps.

**Comment by abhinandanpb** [comment]
hjdr4 thanks for the patch. That is not the right one either. We need the copy from the store to work on the latest ep object . I have the fix. I am going to update the PR with few test cases to ensure corner cases are covered. 
@mvdstam on it .

**Comment by hjdr4** [comment]
addServiceInfoToCluster() calls ep.getNetwork() so it sounded correct to use the same network object accross the whole interface call. I just dig into the code so I'm probably wrong (and I couldn't find any doc on store designs). The good news is that people are working on the subject. I'll wait for the proper patch release. Thank you!

**Comment by mavenugo** [comment]
@hjdr4 as @abhinandanpb suggested, we have a few issues with the store/caching layer and the way endpoints are handled in the layers above the store/cache layer. Hence it is safe to not assume the object references. We do have an open item to address this issue.

**Comment by fcrisciani** [comment]
did we close the discussion about the store?

**Comment by mvdstam** [comment]
@abhinandanpb @fcrisciani @mavenugo @hjdr4 

Hi guys, any news on this? Any chance we could get this merged soon? Thanks!

**Comment by abhinandanpb** [comment]
@mvdstam apologize for the delay. We are testing out few scenarios. There have been some race conditions that have been fixed lately. So in the context of that we are ensuring we don't introduce another issue fixing this.

**Comment by mvdstam** [comment]
@abhinandanpb Thanks for the quick response, I understand. Just checking if this is still on the radar. 😃 

Thanks!

**Comment by fcrisciani** [comment]
@abhinandanpb I was wondering if there is a way to change the container orchestrator to ensure to call the disable service under any condition. If that is possible we can remove the logic from the sbLeave and keep it only in the disable service and also we can ignore the use of the database. WDYT?

**Comment by sirlatrom** [comment]
Is there anything I as an outsider can do to help progress this PR? Since we're running Docker Enterprise Edition, we cannot 'simply' compile dockerd with this PR in it, as we will then not be able to get support. We've already tried to get more attention to this issue through our support plan.

**Comment by sirlatrom** [comment]
PS: From what I can see, the merge conflict is really very simple to resolve, as it's just two lines with independent changes that just happen to be at the same lines.

**Comment by mvdstam** [comment]
I totally agree with @sirlatrom. Again: I'd be happy to provide more information with tests and examples as I've done throughout moby/moby#30321. Hope to see some movement in this PR soon; we really need to be able to perform rolling updates without service interruption.

Ping @abhinandanpb @fcrisciani

**Comment by fcrisciani** [comment]
hey guys, sorry for the delay @abhinandanpb was OOO at the moby summit, I think he will take care of it next week when he will be back

**Comment by andrewhsu** [comment]
@abhinandanpb any update on this?

**Comment by abhinandanpb** [comment]
@andrewhsu Will be working on this PR - this week. We have decided to make changes both in moby/moby and libnetwork.

**Comment by sirlatrom** [comment]
@abhi @andrewhsu That sounds great! Is there an issue and/or a PR in moby/moby that I can track for progress and/or help out with reproduction and the like?

**Comment by mvdstam** [comment]
@andrewhsu @abhi Has this been worked on yet? Can we expect moby/moby#30321 to be fixed with the next release?

**Comment by YarekTyshchenko** [comment]
Can confirm that this fixes downtime issues in swarm deployments for us too. Debian 8.9, patched Docker 17.06

**Comment by anojansivarajah** [comment]
When can we expect this to be fixed?

**Comment by anojansivarajah** [comment]
Hey guys, any idea when this issue will be fixed?

File: sandbox.go [hunk 8]
```diff
@@ -674,24 +674,62 @@ func (sb *sandbox) SetKey(basePath string) error {
...
+			sb.DisableService()
+		}
+	}()
 	for _, ep := range sb.getConnectedEndpoints() {
-		if ep.enableService(true) {
-			if err := ep.addServiceInfoToCluster(sb); err != nil {
-				ep.enableService(false)
-				return fmt.Errorf("could not update state for endpoint %s into cluster: %v", ep.Name(), err)
+		if !ep.isServiceEnabled() {
+			n, err = ep.getNetworkFromStore()
```
Comment (fcrisciani) [inline sandbox.go]: What does the store add here? can we keep the original logic that uses the *Endpoint coming from the map?

File: endpoint.go [hunk 9]
```diff
@@ -759,10 +768,6 @@ func (ep *endpoint) sbLeave(sb *sandbox, force bool, options ...EndpointOption)
 		return err
 	}
 
-	if e := ep.deleteServiceInfoFromCluster(sb, "sbLeave"); e != nil {
```
Comment (fcrisciani) [inline endpoint.go]: without this one will it still work the network disconnect?

File: sandbox.go [hunk 10]
```diff
@@ -674,24 +674,62 @@ func (sb *sandbox) SetKey(basePath string) error {
...
+	defer func() {
+		if len(failedEps) > 0 {
+			err = fmt.Errorf("failed to disable service on sandbox:%s, for endpoints %s", sb.ID(), strings.Join(failedEps, ","))
+		}
+	}()
 	for _, ep := range sb.getConnectedEndpoints() {
-		ep.enableService(false)
+		if ep.isServiceEnabled() {
+			n, err := ep.getNetworkFromStore()
+			if err != nil {
```
Comment (fcrisciani) [inline sandbox.go]: same here, do we need the store data?

File: endpoint.go (hunk 9 above)
Comment (abhi) [reply endpoint.go]: check the corresponding moby/moby PR for this

File: sandbox.go (hunk 8 above)
Comment (abhi) [reply sandbox.go]: store doesnt add anything.  Iam following the convention on of operating on the store object

File: sandbox.go [hunk 11]
```diff
@@ -674,24 +674,41 @@ func (sb *sandbox) SetKey(basePath string) error {
...
+		if ep.isServiceEnabled() {
+			if err := ep.deleteServiceInfoFromCluster(sb, "DisableService"); err != nil {
+				failedEps = append(failedEps, ep.Name())
+				logrus.Warnf("failed update state for endpoint %s into cluster: %v", ep.Name(), err)
+			}
+			ep.disableService()
+		}
 	}
 	logrus.Debugf("DisableService %s DONE", sb.containerID)
 	return nil
```
Comment (fcrisciani) [inline sandbox.go]: return err :)
